File: vk4j/vkxml2java/codegen/enum.py
from ..entity import Enum


def generate_enum(enum: Enum) -> str:
    if len(enum.variants) == 0:
        return f'''package tech.icey.vk4j.enumtype;

public final class {enum.name} {{
}}
'''

    return f'''package tech.icey.vk4j.enumtype;

import tech.icey.panama.annotation.enumtype;

public final class {enum.name} {{
{generate_enum_fields(enum)}

{generate_normal_enum_explain(enum)}
}}
'''


def generate_enum_fields(enum: Enum) -> str:
    lines = []
    for variant in enum.variants:
        lines.append(f'    public static final int {variant.name} = {variant.value};')
    return '\n'.join(lines)


def generate_normal_enum_explain(enum: Enum) -> str:
    return f'''    public static String explain(@enumtype({enum.name}.class) int value) {{
        return switch (value) {{
{'\n'.join(f'            case {variant.name} -> "{variant.name}";' for variant in enum.variants)}
            default -> "Unknown";
        }};
    }}'''

File: codegen/src/main/kotlin/codegen/enum.kt
package codegen

import Enumeration
import appendLn

fun generateEnum(packageName: String, enumeration: Enumeration): String = buildString {
    appendLn("package $packageName.enumtype;")
    appendLn()
    if (enumeration.variants.isEmpty()) {
        appendLn("public final class ${enumeration.name} {")
        appendLn("}")
    }
    else {
        appendLn("import tech.icey.panama.annotation.enumtype;")
        appendLn()
        appendLn("public final class ${enumeration.name} {")
        for (variant in enumeration.variants) {
            appendLn("    public static final int ${variant.name} = ${variant.value};")
        }
        appendLn()
        appendLn("    public static String explain(@enumtype(${enumeration.name}.class) int value) {")
        appendLn("        return switch (value) {")
        for (variant in enumeration.variants) {
            appendLn("            case ${enumeration.name}.${variant.name} -> \"${variant.name}\";")
        }
        appendLn("            default -> \"Unknown ${enumeration.name}: \" + value;")
        appendLn("        };")
        appendLn("    }")
        appendLn("}")
    }
}
